### ISSUE TITLE
Support exclude rules config for cache.buildDependencies

### ISSUE THREAD

**Issue by hiyuki** [webpack/webpack#15466]
## Feature request

**What is the expected behavior?**

Exclude some files/directories from the buildDependencies collected by `FileSystemInfo.resolveBuildDependencies()`

**What is motivation or use case for adding/changing the behavior?**

Sometimes some loader reads config from `projectRoot/package.json` by `require()` (for example `postcss-load-config -> lilconfig`) in this case `projectRoot/package.json` will be collected as a buildDependency by `FileSystemInfo.resolveBuildDependencies()`, and any change of `projectRoot/package.json` (such as project dependency update) will cause the invalidation of the full filesystem cache.

Actually I have not written any postcss config in my package.json file, so I want a exclude rules config to remove `projectRoot/package.json` from the buildDependencies. 

**How should this be implemented in your opinion?**
Support exclude rules config for cache.buildDependencies

**Are you willing to work on this yourself?**
yes

**Comment by vankop** [comment]
@hiyuki this behaviour is expected, since this `package.json` is a build dependency

**Comment by hiyuki** [comment]
But in this case I know I have no postcss config in the `package.json`, but in the `postcss.config.js`, because of `lilconfig` trying to `require` both of them, the `package.json` has been regarded as a build dependency, actually the change of `package.json` has no side effect to the webpack compilation.

**Comment by vankop** [comment]
for webpack it is impossible to determine this, so it adds all to build dependencies.

**Comment by hiyuki** [comment]
> for webpack it is impossible to determine this, so it adds all to build dependencies.

Yes I know so I want a config to customize the build dependencies.

**Comment by alexander-akait** [comment]
You can't do it

**Comment by hiyuki** [comment]
Ok, actually I can pass a load config to `lilconfig` to read `package.json` by `fs.readFile` instead of `require` to avoid this problem. 
Thanks for your replies. @vankop @alexander-akait

**Comment by starmooms** [comment]
In my case, we have to change the version number of `package.json` when we release the project. This leads to cache invalidation. In fact, the dependent module has not changed.

I agree that the cache should be cleared when dependency changes. It would be better to provide some options for more precise control.

For example, some options are provided to control the generation of cache 'hash' of some files?

**Comment by vankop** [comment]
@starmooms not all cache invalidates in your case

**Comment by starmooms** [comment]
But, it will take more than `30s` to re resolve build dependencies...

**Comment by IMalyugin** [comment]
Upvoting the request a lot. Have a similar build, we got a shared monorepository with all the common packages in an enterprise build. Build takes about 10 minutes, if cached it can run 30 seconds.

Unfortunately our common packages are all published on each change, even if there aren't changes in one of the packages.
Webpack build is based on `@scope/conf` package, which contains core config shared throughout all of the projects, once `@scope/conf/package.json` version is raised (which is every build), buildDependencies are mismatched and the entire cache invalidated.

Spent a good couple of days trying to hack into `PackFileCacheStrategy` and `FileSystemInfo`, to allow exclusion of package.json in build dependencies with no luck, good job on isolating it :)

Instead I had to copy/paste the cache section of `WebpackOptionsApply` into a custom plugin to monkeyPatch `PackFileCacheStrategy.createSnapshot` into filtering out `package.json` files. That fixed the issue, but the price is a bit too expensive.

It would be nice to have:
- a simple tapable before createSnapshot method call to insert custom plugin logic (maybe a tapable inside Cache to return an instance of the Strategy?);
- a way to output the list of mismatched buildDependencies paths, so that it's possible to rewrite configuration files avoiding mismatch (possibly another tapable, or a publicly available field inside PackFileCacheStrategy).
